File: SoftwareEngineering/DesignPattern/Builder/typescript/Builder.ts
export class Car {
    public constructor(
        public make: string,
        public model: string,
        public color: string,
        public numDoors: number
    ) { }
}

export interface ICarBuilder {
    color: string;
    numDoors: number;

    getResult(): Car;
}

export class FerrariBuilder implements ICarBuilder {
    public color: string;
    public numDoors: number;

    public getResult(): Car {
        return this.numDoors == 2 ? new Car("Ferrari", "488 Spider", this.color, this.numDoors) : null;
    }
}

export class SportsCarBuildDirector {
    public constructor(private builder: ICarBuilder) { }

    public construct(): void {
        this.builder.color = "Red";
        this.builder.numDoors = 2;
    }
}


const builder = new FerrariBuilder();
const director = new SportsCarBuildDirector(builder);

director.construct();
const myRaceCar = builder.getResult();

File: SoftwareEngineering/DesignPattern/Builder/csharp/Builder.cs
/// <summary>
/// Represents a product created by the builder
/// </summary>
public class Car
{
    public string Make { get; set; }
    public string Model { get; set; }
    public int NumDoors { get; set; }
    public string Colour { get; set; }

    public Car(string make, string model, string colour, int numDoors)
    {
        Make = make;
        Model = model;
        Colour = colour;
        NumDoors = numDoors;
    }
}

/// <summary>
/// The builder abstraction
/// </summary>
public interface ICarBuilder
{
    string Colour { get; set; }
    int NumDoors { get; set; }

    Car GetResult();
}

/// <summary>
/// Concrete builder implementation
/// </summary>
public class FerrariBuilder : ICarBuilder
{
    public string Colour { get; set; }
    public int NumDoors { get; set; }

    public Car GetResult()
    {
        return NumDoors == 2 ? new Car("Ferrari", "488 Spider", Colour, NumDoors) : null;        
    }
}

/// <summary>
/// The director
/// </summary>
public class SportsCarBuildDirector
{
    private ICarBuilder _builder;
    public SportsCarBuildDirector(ICarBuilder builder) 
    {
        _builder = builder;
    }

    public void Construct()
    {
        _builder.Colour = "Red";
        _builder.NumDoors = 2;
    }
}

public class Client
{
    public void DoSomethingWithCars()
    {

        var builder = new FerrariBuilder();
        var director = new SportsCarBuildDirector(builder);

        director.Construct();
        Car myRaceCar = builder.GetResult();
    }
}